R/mash_wrapper.R

```diff
@@ -53,33 +53,40 @@ filter_invalid_summary_stat <- function(dat_list, bhat = NULL, sbhat = NULL, z =
 }
 
 #' @export
-filter_mixture_components <- function(conditions_to_keep, U, w = NULL, w_cutoff = 1e-4) {
-  # Check if colnames of Y is a subset of column names of each element in U
-  U <- lapply(U, function(mat, to_keep) {
-    missing_conditions <- setdiff(conditions_to_keep, colnames(mat))
-    if (length(missing_conditions) > 0) {
-      stop(paste("Condition(s)", paste(missing_conditions, collapse = ", "), "not found in matrix", mat_name))
-    }
-    mat[to_keep, to_keep]
-  }, conditions_to_keep)
-
-  for (mat_name in names(U)) {
-    # remove null component
-    if (all(U[[mat_name]] == 0)) {
-      U[[mat_name]] <- NULL
-      if (!is.null(w)) w <- w[!names(w) %in% mat_name]
-      next
-    }
-    if (!is.null(w)) {
-      if (w[mat_name] < w_cutoff) {
-        w <- w[!names(w) %in% mat_name]
-        U[[mat_name]] <- NULL
-      }
-      next
+filter_mixture_components <- function (conditions_to_keep, U, w = NULL, w_cutoff = 1e-04) {
+    phenos_out <- setdiff(colnames(U[[1]]), conditions_to_keep)
+    U <- lapply(U, function(mat, to_keep) {
+        missing_conditions <- setdiff(conditions_to_keep, colnames(mat))
+        if (length(missing_conditions) > 0) {
+            stop(paste("Condition(s)", paste(missing_conditions, 
+                collapse = ", "), "not found in matrix", mat_name))
+        }
+        mat[to_keep, to_keep]
+    }, conditions_to_keep)
+    
+    for (mat_name in names(U)) {
+        # remove null component
+        if (all(U[[mat_name]] == 0)) {
+            U[[mat_name]] <- NULL
+            if (!is.null(w)) 
+                w <- w[!names(w) %in% mat_name]
+            next
+        }
+        if (!is.null(w)) {
+            if (w[mat_name] < w_cutoff) {
+                w <- w[!names(w) %in% mat_name]
+                U[[mat_name]] <- NULL
+            }
+            next
+        }
     }
-  }
-  message(paste(length(U), "components of matrices remained after filtering. "))
-  return(list(U = U, w = w))
+    # remove the data driven U in U list, and recalculate w
+    U[phenos_out] <- NULL
```

Comment:
this line does not look correct to me. After subsetting `U` (line 58 to 65), if the remaining elements in U are all zeros, as in the case of a singleton matrix removed from the mixture, then we should already set those matrices NULL.

Reply:
The original U is like below:

> 'XtX''tFLASH_default''FLASH_default_1''FLASH_default_2''FLASH_default_3''FLASH_default_4''FLASH_default_5''FLASH_default_6''FLASH_default_7''FLASH_default_8''FLASH_default_9''FLASH_default_10''FLASH_default_11''tFLASH_nonneg''FLASH_nonneg_1''FLASH_nonneg_2''FLASH_nonneg_3''FLASH_nonneg_4''FLASH_nonneg_5''FLASH_nonneg_6''FLASH_nonneg_7''FLASH_nonneg_8''FLASH_nonneg_9''FLASH_nonneg_10''FLASH_nonneg_11''FLASH_nonneg_12''PCA_1''PCA_2''tPCA''identity''Knight_eQTL_brain''MiGA_GFM_eQTL''MiGA_GTS_eQTL''MiGA_SVZ_eQTL''MiGA_THA_eQTL''BM_10_MSBB_eQTL''BM_22_MSBB_eQTL''BM_36_MSBB_eQTL''BM_44_MSBB_eQTL''monocyte_ROSMAP_eQTL''Mic_DeJager_eQTL''Ast_DeJager_eQTL''Oli_DeJager_eQTL''Exc_DeJager_eQTL''Inh_DeJager_eQTL''DLPFC_DeJager_eQTL''PCC_DeJager_eQTL''AC_DeJager_eQTL''Mic_Kellis_eQTL''Ast_Kellis_eQTL''Oli_Kellis_eQTL''OPC_Kellis_eQTL''Exc_Kellis_eQTL''Inh_Kellis_eQTL''Ast_mega_eQTL''Exc_mega_eQTL''Inh_mega_eQTL''Oli_mega_eQTL''STARNET_eQTL_Mac''DLPFC_Bennett_pQTL''OPC_DeJager_eQTL''OPC_mega_eQTL''Ast_10_Kellis_eQTL''Mic_12_Kellis_eQTL''Mic_13_Kellis_eQTL''Mic_mega_eQTL''STARNET_eQTL''equal_effects''simple_het_1''simple_het_2''simple_het_3'

as you can see, there included some data driven U from each context. To my understanding, we should not keep the filter out contexts in this list, so this line is going to remove the U such as  `BM_22_MSBB_eQTL` when we decide to look into ROSMAP data?

Reply:
I know. Still, my reply is "After subsetting U (line 58 to 65), if the remaining elements in U are all zeros, as in the case of a singleton matrix removed from the mixture, then we should already set those matrices NULL." To elaborate, if a matrix is BM_22_MSBB_eQTL it would be non-zero only at the BM_22_MSBB_eQTL entry. Once we subset and removed that context, this entire compoment `BM_22_MSBB_eQTL` will be all zeros and therefore get removed.

Reply:
So, we have another option here: to either check `if all elements are zero` to `if all elements, except for the diagonal, are zero`.

Reply:
I see, this totally makes sense now for what @Chunmingl observed those redundant diagonal matrices! 

you are right, i forgot how it works ... it was using context specific matrices only as initial values into the EM algorithm but the output will actually change the matrices as it sees fit, possibly slightly, like the above. Perhaps just to make it full rank. But this creates an issue for us. 

We cannot simply remove diagonal matrices because signals on the diagonal can be strong and relevant.

If we stick to exactly our pipeline, then perhaps what you did makes the most sense after all, to get things roughly working .... let's do that. My only request then is to add some comment strings in the code to point to this discussion here, for future references.

Reply:
sure thing, will do